drafts/str4d-stratum/draft1.rst

```diff
@@ -0,0 +1,511 @@
+
+Response:
+
+    {"id": 1, "result": ["*SESSION_ID*", "*NONCE_1*"], "error": null} ``\n``
+
+``SESSION_ID`` (str)
+  The session id, for use when resuming (see `Session Resuming`_).
+
+``NONCE_1`` (hex)
+  The first part of the block header nonce (see `Nonce Parts`_).
+
+``mining.authorize()``
+----------------------
+
+A miner MUST authorize a worker in order to submit solutions. A miner MAY
+authorize multiple workers in the same session; this could be for statistical
+purposes on the particular server being used. Details of such purposes are
+outside the scope of this specification.
+
+Request:
+
+    {"id": 2, "method": "mining.authorize", "params": ["*WORKER_NAME*", "*WORKER_PASSWORD*"]} ``\n``
+
+``WORKER_NAME`` (str)
+  The worker name.
+
+``WORKER_PASSWORD`` (str)
+  The worker name.
+
+Response:
+
+    {"id": 2, "result": *AUTHORIZED*, "error": *ERROR*} ``\n``
+
+``AUTHORIZED`` (bool)
+  MUST be ``true`` if authorization succeeded. Per [JSON-RPC-1.0]_, MUST be
+  ``null`` if there was an error.
+
+``ERROR`` (obj)
+  An error object. MUST be ``null`` if authorization succeeded.
+
+  If authorization failed, the server MUST provide an error object describing
+  the reason. See `Error Objects`_ for the object format.
+
+``mining.set_target()``
+-----------------------
+
+Server message:
+
+    {"id": null, "method": "mining.set_target", "params": ["*TARGET*"]} ``\n``
+
+``TARGET`` (hex)
+  The server target for the next received job and all subsequent jobs (until the
+  next time this message is sent). The miner compares proposed block hashes with
+  this target as a 256-bit big-endian integer, and valid blocks MUST NOT have
+  hashes larger than (above) the current target (in accordance with the Zcash
+  network consensus rules [Zcash-Target]_).
+
+  Miners SHOULD NOT submit work above this target. Miners SHOULD validate their
+  solutions before submission (to avoid both unnecessary network traffic and
+  wasted miner time).
+
+  Servers MUST NOT accept submissions above this target for jobs sent after this
+  message. Servers MAY accept submissions above this target for jobs sent before
+  this message, but MUST check them against the previous target.
+
+When displaying the current target in the UI to users, miners MAY convert the
+target to an integer difficulty as used in Bitcoin miners. When doing so, miners
+SHOULD use ``powLimit`` (as defined in ``src/chainparams.cpp``) as the basis for
+conversion.
+
+.. [Zcash-Target] Daira Hopwood, Sean Bowe, Taylor Hornby, Nathan Wilcox.
+  "Difficulty filter". In: *Zcash Protocol Specification*.
+  Version 2016.0-beta-1.5, Section 6.4.2. September 22, 2016.
+  URL: https://github.com/zcash/zips/blob/master/protocol/protocol.pdf
+  (visited on 2016-09-24).
+
+``mining.notify()``
+-------------------
+
+Server message:
+
+    {"id": null, "method": "mining.notify", "params": ["*JOB_ID*", "*VERSION*", "*PREVHASH*", "*MERKLEROOT*", "*RESERVED*", "*TIME*", "*BITS*", *CLEAN_JOBS*]} ``\n``
```

Comment:
The only part of the `mining.notify` params that isn't part of the block header is `JOB_ID` (and for `mining.submit`, `WORKER_NAME` and `JOB_ID`). Everything else is potentially going to differ between different block headers, quite possibly more than the modifications Zcash has made. So I'm not inclined to agree that retrofitting the Zcash modifications into the Bitcoin Stratum format is a good idea, particularly because we may want to change the block header format again in future in a way that doesn't work with this retrofitting.

I was about to say that the only way I would be convinced is if we might in future need to be able to specify a nonce in the coinbase again, and thus need to use a merkle path. But:

1. There should never be a need to do so for the current proof-of-work because of both its relative slowness and the increased size of the nonce.
2. A change to a different proof-of-work that made coinbase nonces necessary would itself require changing the block header version, and the current draft specifies that future versions can have different parameters.